sandbox-backend/src/main/java/com/exadel/sandbox/team5/service/impl/DiscountServiceImpl.java

```diff
@@ -62,4 +67,39 @@ public DiscountDto update(DiscountDto discount) {
     public void delete(Long id) {
         discountDAO.deleteById(id);
     }
+
+    @Override
+    public Page<DiscountDto> getByCriteria(DiscountSearchCriteria searchCriteria) {
+        String searchText = getWildcard(searchCriteria.getSearchText());
+        List<Discount> result;
+        List<Long> discountId = new ArrayList<>();
+        if (searchCriteria.getTags() == null || searchCriteria.getTags().isEmpty()) {
+            result = discountDAO.getByCriteria(searchText, searchCriteria.getRate());
+        } else {
+            result = discountDAO.getByCriteriaWithTags(searchText,
+                    searchCriteria.getTags(), searchCriteria.getRate());
+        }
+        result.forEach(x -> discountId.add(x.getId()));
+        List<RateSetter> rateList = reviewDAO.getRateByDiscountId(discountId);
+        List<DiscountDto> discountDTOs = mapper.mapAll(result, DiscountDto.class);
+        discountDTOs = setRate(rateList, discountDTOs);
+        return new PageImpl<>(discountDTOs, searchCriteria.getPageRequest(), discountDTOs.size());
+    }
+
+    private String getWildcard(String text) {
```

Comment:
Let's extract it to `QueryUtils` class as static method

Reply:
done